plugins/channel-safely/channel-safely-plugin.cpp

```diff
@@ -0,0 +1,666 @@
+        -> Unsafe: insta-dig & delete job (presumes the job is only accessible from directly on the tile)
+        -> Unsafe: set designation to Marker Mode (no insta-digging)
+ -> check tile occupancy:
+    -> HasUnit: check if a fall would be safe:
+        -> Safe: do nothing, let them fall
+        -> Unsafe: remove worker for 1 tick (test if this "pauses" or cancels the job)
+        -> Unsafe: Add feature to teleport unit?
+
+ Job completed event:
+ -> validate job type (channel)
+ -> verify completion:
+    -> IsOpenSpace: mark done
+    -> IsOpenSpace: manage tile below
+    -> NotOpenSpace: check for designation
+        -> HasDesignation: do nothing
+        -> NoDesignation: mark done (erases from group)
+        -> NoDesignation: manage tile below
+*/
+
+#include <plugin.h>
+#include <inlines.h>
+#include <channel-manager.h>
+#include <tile-cache.h>
+
+#include <Debug.h>
+#include <LuaTools.h>
+#include <LuaWrapper.h>
+#include <PluginManager.h>
+#include <modules/EventManager.h>
+#include <modules/Units.h>
+#include <df/world.h>
+#include <df/report.h>
+#include <df/tile_traffic.h>
+#include <df/block_square_event_designation_priorityst.h>
+
+#include <cinttypes>
+#include <unordered_map>
+#include <unordered_set>
+
+// Debugging
+namespace DFHack {
+    DBG_DECLARE(channelsafely, plugin, DebugCategory::LINFO);
+    DBG_DECLARE(channelsafely, monitor, DebugCategory::LERROR);
+    DBG_DECLARE(channelsafely, manager, DebugCategory::LERROR);
+    DBG_DECLARE(channelsafely, groups, DebugCategory::LERROR);
+    DBG_DECLARE(channelsafely, jobs, DebugCategory::LERROR);
+}
+
+DFHACK_PLUGIN("channel-safely");
+DFHACK_PLUGIN_IS_ENABLED(enabled);
+REQUIRE_GLOBAL(world);
+
+namespace EM = EventManager;
+using namespace DFHack;
+using namespace EM::EventType;
+
+int32_t mapx, mapy, mapz;
+Configuration config;
+PersistentDataItem psetting;
+PersistentDataItem pfeature;
+const std::string FCONFIG_KEY = std::string(plugin_name) + "/feature";
+const std::string SCONFIG_KEY = std::string(plugin_name) + "/setting";
+
+enum FeatureConfigData {
+    VISION,
+    MONITOR,
+    RESURRECT,
+    INSTADIG
+};
+
+enum SettingConfigData {
+    REFRESH_RATE,
+    MONITOR_RATE,
+    IGNORE_THRESH,
+    FALL_THRESH
+};
+
+// dig-now.cpp
+df::coord simulate_fall(const df::coord &pos) {
+    df::coord resting_pos(pos);
+
+    while (Maps::ensureTileBlock(resting_pos)) {
+        df::tiletype tt = *Maps::getTileType(resting_pos);
+        df::tiletype_shape_basic basic_shape = tileShapeBasic(tileShape(tt));
+        if (isWalkable(tt) && basic_shape != df::tiletype_shape_basic::Open)
+            break;
+        --resting_pos.z;
+    }
+
+    return resting_pos;
+}
+
+df::coord simulate_area_fall(const df::coord &pos) {
+    df::coord neighbours[8]{};
+    get_neighbours(pos, neighbours);
+    df::coord lowest = simulate_fall(pos);
+    for (auto p : neighbours) {
+        if (p.z < lowest.z) {
+            lowest = p;
+        }
+    }
```

Comment:
if you aren't simulating falling for the neighbors, when will this for loop ever change the value for lowest?

Reply:
oops, brain must have been faster than my fingers